File: src/mvc/cpp/emoji_picker/emoji_picker_tree.cpp
#include "emoji_picker_tree.h"
#include "emoji_picker_delegate.h"
#include <QHeaderView>

EmojiPickerItem::EmojiPickerItem(const EmojiCategory &category, int columnCount)
    : QStandardItem(), itemCategory(category)
{
   setColumnCount(columnCount);
   setCategory(category);
}


void EmojiPickerItem::setCategory(const EmojiCategory &category)
{
    itemCategory = category;
    setText(category.name);
    for (int i = 0; i * columnCount() < category.emojis.size(); i++) {
        QList<QStandardItem *> items;
        for (auto j = 0; j < columnCount(); ++j) {
            auto next_index = j + i * columnCount();
            if (next_index >= category.emojis.size()) {
                break;
            }
            auto icon = QIcon(category.emojis[next_index].iconPath);
            items.append(new QStandardItem(icon, ""));
        }
        appendRow(items);
    }
}

void EmojiPickerItem::setColumnCount(int count)
{
    QStandardItem::setColumnCount(count);
    setCategory(itemCategory);
}


const EmojiCategory &EmojiPickerItem::category() const
{
    return itemCategory;
}

EmojiPickerModel::EmojiPickerModel(QObject *parent)
    : QStandardItemModel(parent)
{
}


EmojiPickerModel::~EmojiPickerModel()
{
}


void EmojiPickerModel::appendRow(const EmojiCategory &category)
{
    QStandardItemModel::appendRow(new EmojiPickerItem(category, columnCount()));
}


void EmojiPickerModel::appendRow(const QList<EmojiCategory> &categories)
{
    for (int i = 0; i < categories.size(); i++) {
        appendRow(categories[i]);
    }
}


void EmojiPickerModel::clear()
{
    removeRows(0, rowCount());
}

void EmojiPickerModel::setColumnCount(int count)
{
   QStandardItemModel:: setColumnCount(count);
   for (auto item : findItems("", Qt::MatchContains)) {
        item->setColumnCount(count);
 }
}


EmojiPickerTree::EmojiPickerTree(QWidget *parent)
    : QTreeView(parent)
{
    setHeaderHidden(true);
    setItemsExpandable(false);
    setStyleSheet("QTreeView::branch {  border-image: url(none.png); }");
    setEditTriggers(QAbstractItemView::NoEditTriggers);
    setHorizontalScrollBarPolicy(Qt::ScrollBarAlwaysOff);
    setSelectionBehavior(QAbstractItemView::SelectItems);
    setItemDelegate(new EmojiPickerDelegate(this));
    header()->setStretchLastSection(false);
}


EmojiPickerTree::~EmojiPickerTree()
{
}


void EmojiPickerTree::setModel(EmojiPickerModel* model)
{
    QTreeView::setModel(model);
    expandAll();
    setIndentation(3);
    header()->setSectionResizeMode(0, QHeaderView::Fixed);
    for (auto column = 0; column < model->columnCount(); column++) {
        header()->setSectionResizeMode(column, QHeaderView::Fixed);
        header()->resizeSection(column, 50);
    }
    for (auto row = 0; row < model->rowCount(); row++) {
        setFirstColumnSpanned(row, QModelIndex(), true);
    }
}


EmojiPickerModel* EmojiPickerTree::model()
{
    return static_cast<EmojiPickerModel*>(QTreeView::model());
}

File: src/mvc/python/emoji_picker/emoji_picker_tree.py
from PyQt6.QtCore import Qt, QModelIndex
from PyQt6.QtGui import QIcon, QStandardItem, QStandardItemModel
from PyQt6.QtWidgets import QTreeView, QAbstractItemView, QHeaderView

from emoji_picker_delegate import EmojiPickerDelegate


class EmojiPickerItem(QStandardItem):
    def __init__(self, category, column_count):
        super().__init__()
        self.itemCategory = category
        self.setColumnCount(column_count)
        self.setCategory(category)

    def setCategory(self, category):
        self.itemCategory = category
        self.setText(category.name)
        for i in range(0, len(category.emojis) // self.columnCount()):
            items = []
            for j in range(0, self.columnCount()):
                next_index = j + i * self.columnCount()
                if next_index >= len(category.emojis):
                    break
                icon = QIcon(category.emojis[next_index].iconPath)
                items.append(QStandardItem(icon, ""))
            self.appendRow(items)
    
    def setColumnCount(self, count):
        QStandardItem.setColumnCount(self, count)
        self.setCategory(self.itemCategory)

    def category(self):
        return self.itemCategory

    
class EmojiPickerModel(QStandardItemModel):
    def __init__(self, parent):
        super().__init__(parent)
        
    def appendRow(self, category):
        QStandardItemModel.appendRow(self, EmojiPickerItem(category, self.columnCount()))

    def clear(self):
        QStandardItemModel.removeRows(self, 0, self.rowCount())

    def setColumnCount(self, count):
        QStandardItemModel.setColumnCount(self, count)
        for item in self.findItems("", Qt.MatchFlag.MatchContains):
            item.setColumnCount(count)


class EmojiPickerTree(QTreeView):
    def __init__(self, parent):
        super().__init__(parent)
        self.setHeaderHidden(True)
        self.setItemsExpandable(False)
        self.setStyleSheet("QTreeView::branch {  border-image: url(none.png); }")
        self.setEditTriggers(QAbstractItemView.EditTrigger.NoEditTriggers)
        self.setHorizontalScrollBarPolicy(Qt.ScrollBarPolicy.ScrollBarAlwaysOff)
        self.setSelectionBehavior(QAbstractItemView.SelectionBehavior.SelectItems)
        self.setItemDelegate(EmojiPickerDelegate(self))
        self.header().setStretchLastSection(False)
    
    def setModel(self, model):
        super().setModel(model)
        self.expandAll()
        self.setIndentation(3)
        self.header().setSectionResizeMode(0, QHeaderView.ResizeMode.Fixed)
        for column in range(0, model.columnCount()):
            self.header().setSectionResizeMode(column, QHeaderView.ResizeMode.Fixed)
            self.header().resizeSection(column, 50)
        for row in range(0, model.rowCount()):
            self.setFirstColumnSpanned(row, QModelIndex(), True)

    def model(self):
        return super().model()
